scipy/linalg/_cythonized_array_utils.pyx

```diff
@@ -0,0 +1,378 @@
+
+    Returns
+    -------
+    lu: tuple
+        2-tuple of ints indicating the lower and upper bandwith. A zero
+        denotes no sub- or super-diagonal on that side (triangular), and,
+        say for N rows (N-1) means that side is full. Same example applies
+        to the upper triangular part with (M-1).
+
+    Notes
+    -----
+    This helper function simply runs over the array looking for the nonzero
+    entries whether there exists a banded structure in the array or not. Hence,
+    the performance depends on the density of nonzero entries and also
+    memory-layout. Fortran- or C- contiguous arrays are handled best and
+    otherwise suffers from extra random memory access cost.
+
+    The strategy is to look for only untested band elements in the upper
+    and lower triangular parts separately; depending on the memory layout
+    we scan row-wise or column-wise. Moreover, say we are scanning rows
+    and in the 6th row, 4th entry is nonzero then, on the succeeding rows
+    the horizontal search is done only up to that band entries since we know
+    that band is occupied. Therefore, a completely dense matrix scan cost is
+    in the the order of n.
+
+    Raises
+    ------
+    TypeError
+        If the dtype of the array is not supported, in particular, NumPy
+        float16, float128 and complex256 dtypes.
+    ValueError
+        If the input array is not a 2D NumPy array.
+
+    Examples
+    --------
+    >>> from scipy.linalg import get_array_bandwidth
+    >>> A = np.array([[3., 0., 0., 0., 0.],
+                      [0., 4., 0., 0., 0.],
+                      [0., 0., 5., 1., 0.],
+                      [8., 0., 0., 6., 2.],
+                      [0., 9., 0., 0., 7.]])
+    >>> get_array_bandwidth(A)
+    (3, 1)
+
+    """
+    if a.size == 0:
+        return (0, 0)
+
+    if a.ndim != 2:
+        raise ValueError('Input array must be a 2D NumPy array.')
+
+    if a.flags['C_CONTIGUOUS']:
+        l, u = get_array_bandwidth_c(a)
+    elif a.flags['F_CONTIGUOUS']:
+        u, l = get_array_bandwidth_c(a.T)
+    else:
+        l, u = get_array_bandwidth_noncontig(a)
+
+    return l, u
+
+
+@cython.initializedcheck(False)
+def get_array_bandwidth_c(np_numeric_t[:, ::1]A):
+    cdef int l, u
+    with nogil:
+        l, u = band_check_internal_c(A)
+    return l, u
+
+
+@cython.initializedcheck(False)
+def get_array_bandwidth_noncontig(np_numeric_t[:, :]A):
+    cdef int l, u
+    with nogil:
+        l, u = band_check_internal_noncontig(A)
+    return l, u
+
+
+@cython.initializedcheck(False)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef inline (int, int) band_check_internal_c(np_numeric_t[:, ::1]A) nogil:
+    cdef Py_ssize_t n = A.shape[0], m = A.shape[1]
+    cdef Py_ssize_t lower_band = 0, upper_band = 0, r, c
+    cdef np_numeric_t zero = 0
+
+    for r in xrange(n):
+        # Only bother if outside the existing band:
+        for c in xrange(min(r-lower_band, m - 1)):
```

Comment:
I think this can be improved. Consider the case of a totally full array.  It needs only to check two elements, the `(0, m-1)` and ` (n-1, 0)` and then return, making the whole thing O(1).   As this is currently written, the `lower_band` check has to evaluate  a minimum of `n` elements because it starts with `r = 0` instead of `r = n-1` making the whole thing O(n).

Reply:
That's true but the some of the speed actually comes from using the contiguous memory layout. If we start from bottom then upper band has to wait until the end with comparable speed. Looking at opposite corners can be done already at Python level anyways. O(n) vs O(1) is quite comparable for the most use cases that I think this will be used for but I'll try bottom up scan just to make sure.